src/gc.c

```diff
@@ -3036,6 +3213,118 @@ JL_DLLEXPORT jl_value_t *jl_gc_alloc_3w(void)
+{
+    p = (char *) p - 1;
+    jl_gc_pagemeta_t *meta = page_metadata(p);
+    if (meta && meta->ages) {
+        char *page = gc_page_data(p);
+        // offset within page.
+        size_t off = (char *)p - page;
+        if (off < GC_PAGE_OFFSET)
+            return NULL;
+        // offset within object
+        size_t off2 = (off - GC_PAGE_OFFSET);
+        size_t osize = meta->osize;
+        off2 %= osize;
+        if (off - off2 + osize > GC_PAGE_SZ)
+            return NULL;
+        jl_taggedvalue_t *cell = (jl_taggedvalue_t *)((char *)p - off2);
+        // We have to distinguish between three cases:
+        // 1. We are on a page where every cell is allocated.
+        // 2. We are on a page where objects are currently bump-allocated
+        //    from the corresponding pool->newpages list.
+        // 3. We are on a page with a freelist that is used for object
+        //    allocation.
+        if (meta->nfree == 0) {
+            // case 1: full page; `cell` must be an object
+            goto valid_object;
+        }
+        jl_gc_pool_t *pool =
+            jl_all_tls_states[meta->thread_n]->heap.norm_pools +
+            meta->pool_n;
+        if (meta->fl_begin_offset == (uint16_t) -1) {
+            // case 2: this is a page on the newpages list
+            jl_taggedvalue_t *newpages = pool->newpages;
+            // Check if the page is being allocated from via newpages
+            if (!newpages)
+                return NULL;
+            char *data = gc_page_data(newpages);
+            if (data != meta->data) {
+                // Pages on newpages form a linked list where only the
+                // first one is allocated from (see reset_page()).
+                // All other pages are empty.
+                return NULL;
+            }
+            // This is the first page on the newpages list, where objects
+            // are allocated from.
+            if ((char *)cell >= (char *)newpages) // past allocation pointer
+                return NULL;
+            goto valid_object;
+        }
+        // case 3: this is a page with a freelist
+        // marked or old objects can't be on the freelist
+        if (cell->bits.gc)
+            goto valid_object;
+        // When allocating from a freelist, three subcases are possible:
+        // * The freelist of a page has been exhausted; this was handled
+        //   under case 1, as nfree == 0.
+        // * The freelist of the page has not been used, and the age bits
+        //   reflect whether a cell is on the freelist or an object.
+        // * The freelist is currently being allocated from. In this case,
+        //   pool->freelist will point to the current page; any cell with
+        //   a lower address will be an allocated object, and for cells
+        //   with the same or a higher address, the corresponding age
+        //   bit will reflect whether it's on the freelist.
+        // Age bits are set in sweep_page() and are 0 for freelist
```

Comment:
As I said, for this to work reliably you still need to disable the age reset in the finalizer marking. There should be only a single place where `mark_reset_age` is set to 1 so just disable that. Also, in general, it's still good to have an API for enabling support for conservative scanning. That function should also return whether the conservative scanning support was enabled and you should document that if the setup doesn't guarantee that the support was enabled early enough, the caller should do a full GC afterwards.

Reply:
Okay, I need to unpack that, as I think I simply misunderstood you before (thinking you were referring to sweeping rather than finalizers). I take it now that `mark_reset_age = 1` should be disabled for conservative scanning only?

I had removed the function for enabling conservative scanning because it had become a no-op; but it does make sense to have one regardless in case future changes necessitate having one. And if I understand you correctly, the `mark_reset_age` change is for the conservative case alone and should be left in place for the normal case, so we'd have to put code in there again anyway.

Reply:
> I take it now that mark_reset_age = 1 should be disabled for conservative scanning only?

Yes. Sorry if it wasn't clear.

> but it does make sense to have one regardless in case future changes necessitate having one.

That's exactly what I'm saying.

Reply:
> but it does make sense to have one regardless in case future changes necessitate having one.

And in particular, the behavior of "if the function returns 0 then the caller should do a full GC if it wasn't called right after initialization" should be a behavior that is reasonably easy to maintain.

Reply:
So, does that mean `mark_reset_age` is "just" an optimization, which is thus safe to disable for us? I wonder how much impact disabling it has, esp. compared to our previous version of this PR?

Reply:
Okay. Will do.

Reply:
Yes it's just an optimization. It could delay the freeing of old finalized objects but should not have any other effects.

Reply:
@yuyichao Thanks for the explanation, and of course all the feedback

@rbehrends Perhaps we could extend the comment at the declaration of `mark_reset_age` to state that as part of this PR, so that future people looking at this will have an easier time figuring it out :-)

Reply:
That's fine too. There shouldn't be a issue though since as long as you make sure you don't start doing things that requires conservative scanning before running the GC, the race is benign. Not requiring a external GC does make the API simpler so I'm fine with that. I brought it up only in case you want to reduce the startup delay a little bit.